packages/store/src/atoms/createStoreAtoms.ts

```diff
@@ -0,0 +1,243 @@
+import { atomFamily, loadable } from 'jotai/utils';
+import { createStore as createJotaiStore, atom } from 'jotai/vanilla';
+import type { InterpreterFrom, StateFrom } from 'xstate';
+import { toObserver } from 'xstate';
+
+import type {
+  MachinesObj,
+  Listener,
+  StateItem,
+  Service,
+  StoreServiceObj,
+  StateListener,
+  AddMachineInput,
+} from '../types';
+import { updateService, waitFor } from '../utils';
+
+import { atomWithMachine } from './atomWithMachine';
+import { atomWithSubcription } from './atomWithSubscription';
+
+export type JotaiStore = ReturnType<typeof createJotaiStore>;
+export type CreateStoreAtomsReturn<T extends MachinesObj = MachinesObj> =
+  ReturnType<typeof createStoreAtoms<T>>;
+
+/**
+ * Creates atoms to store machines.
+ * @param id The id of the store.
+ * @returns An object with atoms to store machines.
+ *   - `store`: The global store.
+ *   - `keysAtom`: An atom to get the keys of the machines.
+ *   - `machinesAtom`: An atom to store the machines.
+ *   - `serviceAtom`: An atom family to get an service from a machine.
+ *   - `servicesAtom`: An atom to get all the services.
+ *   - `stateAtom`: An atom family to get the state of a machine.
+ *   - `waitForAtom`: An atom family to wait for a machine to reach a state.
+ *   - `onStateChangeAtom`: An atom family to listen to state changes.
+ *   - `onStoreStartAtom`: An atom to listen to the store start.
+ */
+export function createStoreAtoms<T extends MachinesObj = MachinesObj>(
+  id: string
+) {
+  type Obj = {
+    [K in keyof T]: ReturnType<
+      typeof atomWithMachine<T[K], InterpreterFrom<T[K]>>
+    >;
+  };
+  const store = createGlobalStore(id);
+
+  /**
+   * An atom to store the machines and their services.
+   * @returns An atom to store the machines.
+   * @example
+   * store.set(machinesAtom, {
+   *   key: 'conter',
+   *   getMachine: counterMachine,
+   *   getOptions: {},
+   *   hasStorage: false
+   * })
+   */
+  const machinesObjAtom = atom<Obj>({} as Obj);
+  const machinesAtom = atom(
+    (get) => get(machinesObjAtom),
+    (get, set, input: AddMachineInput<T, keyof T>) => {
+      const curr = get(machinesObjAtom);
+      const machineAtom = atomWithMachine(input);
+      const machinesObj = { ...curr, [input.key]: machineAtom };
+      set(machinesObjAtom, machinesObj);
+      return machinesObj as Obj;
+    }
```

Comment:
Improve the code here by splitting it set / get into a function and passing here by name, it will help to understand what each part does.

Reply:
This could be done for every new atom been created on this file.

Reply:
And for every callback function

Reply:
I create a file for each atom factory we have inside the store, about the callbacks, which will be kinda unnecessary and we'll lose a lot of type inference by doing this. It's not hard to understand, it's the way Jotai works. An atom can have this signature:

```tsx
// primitive atom
function atom<Value>(initialValue: Value): PrimitiveAtom<Value>

// read-only atom
function atom<Value>(read: (get: Getter) => Value | Promise<Value>): Atom<Value>

// writable derived atom
function atom<Value, Update>(
  read: (get: Getter) => Value | Promise<Value>,
  write: (get: Getter, set: Setter, update: Update) => void | Promise<void>
): WritableAtom<Value, Update>

// write-only derived atom
function atom<Value, Update>(
  read: Value,
  write: (get: Getter, set: Setter, update: Update) => void | Promise<void>
): WritableAtom<Value, Update>
```

Basically, when an atom has just the first callback, it's just a read-only atom, when both it's a writable atom (can read, and write).